docs/policy/memtier.md

```diff
@@ -1,70 +1,314 @@
+decide if the workload could potentially benefit from extra resource allocation
+optimizations. Unless configured differently, containers fulfilling certain
+corresponding criteria are considered eligible for these optimizations. This
+will be reflected in the assigned resources whenever that is possible at the
+time the container's creation / resource allocation request hits the policy.
+
+The set of these extra optimizations consist of
+
+  - assignment of `kube-reserved` CPUs
+  - assignment of exclusively allocated CPU cores
+  - usage of kernel-isolated CPU cores (for exclusive allocation)
+
+The policy uses a combination of the QoS class and the resource requirements of
+the container to decide if any of these extra allocation preferences should be
+applied. Containers are divided into five groups, with each group having a
+slightly different set of criteria for eligibility.
+
+  - `kube-system` group
+    * all containers in the `kube-system` namespace
+  - `low-priority` group
+    * containers in the `BestEffort or Burstable` QoS class
+  - `sub-core` group
+    * Guaranteed QoS class containers with `CPU request < 1 CPU`
+  - `mixed` group
+    * Guaranteed QoS class containers with `1 <= CPU request < 2`
+  - `multi-core` group
+    * Guaranteed QoS class containers with `CPU request >= 2`
+
+The eligibility rules for extra optimization are slightly different among these
+groups.
+
+  - `kube-system`
+    * not eligible for extra optimizations
+    * eligible to run on `kube-reserved` CPU cores
+    * always run on shared CPU cores
+  - `low-priority`
+    * not eligible for extra optimizations
+    * always run on shared CPU cores
+  - `sub-core`
+    * not eligible for extra optimizations
+    * always run on shared CPU cores
+  - `mixed`
+    * by default eligible for exclusive and isolated allocation
+    * not eligible for either if `PreferSharedCPUs` is set to true
+    * not eligible for either if annotated to opt out from exclusive allocation
+    * not eligible for isolated allocation if annotated to opt out
+  - `multi-core`
+    * CPU request fractional (`(CPU request % 1000 milli-CPU) != CPU request`):
+      - by default not eligible for extra optimizations
+      - eligible for exclusive and isolated allocation if annotated to opt in
+    * CPU request not fractional:
+      - by default eligible for exclusive allocation
+      - by default not eligible for isolated allocation
+      - not eligible for exclusive allocation if annotated to opt out
+      - eligible for isolated allocation if annotated to opt in
+
+Eligibility for kube-reserved CPU core allocation should always be possible to
+honor. If this is not the case, it is probably due to an incorrect configuration
+which underdeclares `ReservedResources`. In that case, ordinary shared CPU cores
+will be used instead of kube-reserved ones.
+
+Eligibility for exclusive CPU allocation should always be possible to honor. If
```

Comment:
"Eligibility for exclusive CPU allocation should always be possible to honor. ..."

I wouldn't bother to mention the possibility of a bug in the algorithm. There shouldn't be one, and on the other hand, if there is, it could break any other claim as well, not just this.

However, a configuration error like too large ReservedResources cpuset could cause problems. That takes CPUs from real pools without kube scheduler knowing about it.

Reply:
Removed reference to possible bugs in resource accounting.